proposals/2023/03-15_init-step.md

```diff
@@ -0,0 +1,341 @@
+# Add `InitStep` resource
+
+<!--
+The name of this markdown file should:
+
+1. Short and contain no more then 30 characters
+
+2. Contain the date of submission in MM-DD format
+
+3. Clearly state what the proposal is being submitted for
+-->
+
+| Key           | Value              |
+| :-----------: | :----------------: |
+| **Author(s)** | Jacob Floyd        |
+| **Reviewers** |                    |
+| **Date**      | March 15th, 2023   |
+| **Status**    | Waiting for Review |
+
+<!--
+If you're already working with someone, please add them to the proper author/reviewer category.
+
+If not, please leave the reviewer category empty and someone from the Vela team will assign it to themself.
+
+Here is a brief explanation of the different proposal statuses:
+
+1. Reviewed: The proposal is currently under review or has been reviewed.
+
+2. Accepted: The proposal has been accepted and is ready for implementation.
+
+3. In Progress: An accepted proposal is being implemented by actual work.
+
+NOTE: The design is subject to change during this phase.
+
+4. Cancelled: While or before implementation the proposal was cancelled.
+
+NOTE: This can happen for a multitude of reasons.
+
+5. Complete: This feature/change is implemented.
+-->
+
+## Background
+
+<!--
+This section is intended to describe the new feature, redesign or refactor.
+-->
+
+**Please provide a summary of the new feature, redesign or refactor:**
+
+<!--
+Provide your description here.
+-->
+
+Replace the injected "init" step (a pseudo-container) with `InitStep`, a new resource.
+
+Each `Build` will have one or more `InitSteps` that logically group related parts of the log data.
+
+An `InitStep` is a report by a "reporter" about a discrete part of the build setup. The "reporter" would be a logical part of the stack, like the "Pipeline Compiler" or the Runtime. For example, we could have these reporters be discrete initsteps (Reporter: name of the `InitStep`):
+- Pipeline Compiler: report info/debug logs
```

Comment:
Is there an open ticket or request for this information?

Trying to understand the value behind providing this information.

Reply:
No open ticket. I asked in slack as I was trying to figure out what to do about the "init" step.

https://gophers.slack.com/archives/CNRRKE8KY/p1677684779756489?thread_ts=1677684670.251639&cid=CNRRKE8KY

I asked about the "purpose of the injected init step".

@kneal replied with (he mentioned the compiler, emphasis mine):

> The high level idea was to record the ephemeral setup to enable pipeline users to troubleshoot builds without admin level intervention.
>
> Historically there was all sorts of things talked about being added in there. **Compile info**, runtime setup, image pulls, secret setup, etc
>
> Environment injection info was another thing I was personally interested in looking at adding too. It's been a bunch of random things honestly.
>
> But the idea is just provide as much info to the user as possible to self solve without admins and give them a understanding of what's going on under the hood.
>
> ...
>
> Yeah, the init idea in general was built pretty MVP I think. There's a lot of angles for improvement. Even visually honestly, reading that much output isn't super useful. It would be great to have the details from the commands collapsable like Actions does in their setup info.,

Me (@cognifloyd):

> So, I guess to make the PR flow more manageable we'd want to actually add support in reverse order, starting with the UI/CLI where they allow for an optional Init field for reporting all that, and then build backwards to actually add the backend support.

@kneal mentioned the compiler here (my emphasis):

> It doesn't necessarily have to be reverse order either. You could setup the **compiler** and worker to use a new init layer. Get all that in main and then go back and rip things out once the frontend layer is ready.
>
> We'd just be tracking twice but I'm guessing the performance hit is pretty negligible
>
> I think compatibility is more important than that tiny performance loss for a minor period of time.

And then @kneal mentioned the compiler again here (in discussion about moving the init step to some top level field on the Build struct; emphasis is original, not mine):

> If you have init get added in at the pipeline.Build struct layer. You _might_ actually be able to add a whole shit ton of info to that field starting from the compiler layer ...
>
> Adding more info, is likely a double edge sword tho until the output is more easily consumable.

-----------

So, I focused on the logs that could be added from the compiler because @kneal mentioned it several times. It's apparently a blind spot where users have to ask Vela admins for help diagnosing issues. That seems reasonable to me, but it only makes sense when there is a build that such logs can be associated with.